keps/sig-network/1669-proxy-terminating-endpoints/README.md

```diff
@@ -0,0 +1,378 @@
+has not probed kube-proxy's healthCheckNodePort yet. In this event, there is traffic loss because the load balancer is routing traffic to a node where the proxy rules will blackhole
+the traffic due to a lack of local endpoints. The likihood of this traffic loss is impacted by two factors: the number of local endpoints on the node and the interval between health checks
+from the load balancer. The worse case scenario is a node with 1 local endpoint and a load balancer with a long health check interval.
+
+Currently there are several workarounds that users can leverage:
+* Use Kubernetes scheduling/deployment features such that a node would never only have terminating endpoints. For example, always scheduling two pods on a node and only allowing 1 pod to update at a time.
+* Reducing the load balancer health check interval to very small time windows. This may not alwyays be possible based on the load balancer implementation.
+* Use a preStop hook in the Pod to delay the time between a Pod terminating and the process receiving SIGTERM.
+
+While some of these solutions help, there's more that Kubernetes can do to handle this complexity for users.
+
+### Goals
+
+* Reduce potential traffic loss from kube-proxy that occurs on rolling updates because trafffic is sent to Pods that are terminating.
+
+### Non-Goals
+
+* Changing the behavior of how pods terminate.
+* Handling terminating endpoints for other consumers of the EndpointSlice API, such as ingress controllers or external load balancers.
+
+## Proposal
+
+This KEP proposes that if all endpoints for a given Service scoped to its traffic policy are terminating (i.e. pod.DeletionTimestamp != nil), then all traffic should be sent to
+terminating Pods that are still Ready. Note that the EndpointSlice API introduced a new condition called "Serving" which is semantically equivalent to "Ready" except that the Ready condition
+must always be "False" for terminating pods for compatibility reasons. For consumers of the EndpointSLice API that want to route traffic strictly based on a Pod's readiness ignoring
+it's terminating state, they should be reading the Serving condition going forward. Below are some examples to help illustrate the proposed behavior:
+
+### Example: only some endpoints terminating when traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and some endpoints are terminating, all traffic should be routed to the ready endpoints that are not terminating..
+
+### Example: only some endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and some endpoints are terminating within a single node, traffic should be routed to ready endpoints on that node that are not terminating.
+
+### Example: all endpoints terminating and traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and all endpoints are terminating, then traffic should be routed to any terminating endpoint that is ready.
```

Comment:
> Why would someone want this semantic? In the "Local" case, a single node might have only terminating endpoints because of a rolling update. But in the "Cluster" case, if the entire cluster has only terminating endpoints, that suggests that the service is being shut down entirely.

I agree it's less useful for Cluster, but you can still result in scenarios where all endpoints are terminating. For example, all pods might be on the same node that is draining or a 1 pod statefulset, etc.

Reply:
my main concern here is if this needs to add additional logic that we have to support and increase the bug surface area... the use case seems very rare ... but if this add new branches in the code I think we should evaluate the cost

Reply:
> if you have a 1 pod service, then you presumably don't expect it to be "high-availability" in the face of updates / node drains / etc.

Sure, this is just one example where this can happen though. It is also more likely to occur on smaller clusters (like 1-3 nodes)